File: runtime-go/internal/encoder/syslog.go
package encoder

import (
	"encoding/json"
	"fmt"
	"time"

	"github.com/linxinhong/eventweave/runtime-go/internal/event"
)

func syslogPriority(ev event.Event) int {
	facility, _ := ev.Attributes["syslog_facility"].(int)
	if facility == 0 {
		facility = 16
	}
	severity, _ := ev.Attributes["syslog_severity"].(int)
	if severity == 0 {
		severity = 6
	}
	return facility*8 + severity
}

func syslogHostname(ev event.Event) string {
	if h, ok := ev.Attributes["hostname"].(string); ok && h != "" {
		return h
	}
	return ev.SourceID
}

func syslogTag(ev event.Event) string {
	if t, ok := ev.Attributes["syslog_tag"].(string); ok && t != "" {
		return t
	}
	return ev.SourceID
}

func syslogMessage(ev event.Event) string {
	if m, ok := ev.Attributes["message"].(string); ok && m != "" {
		return m
	}
	b, _ := json.Marshal(ev.Attributes)
	return string(b)
}

// RFC3164 encodes events as RFC3164 syslog messages.
type RFC3164 struct{}

func init() {
	Register("syslog-rfc3164", RFC3164{})
}

// Name returns the encoder name.
func (RFC3164) Name() string { return "syslog-rfc3164" }

// ContentType returns the MIME type of the encoded output.
func (RFC3164) ContentType() string { return "text/plain" }

// Encode formats the event as an RFC3164 message.
func (RFC3164) Encode(ev event.Event) ([]byte, error) {
	priority := syslogPriority(ev)
	timestamp := ev.EventTime.Format(time.Stamp)
	hostname := syslogHostname(ev)
	tag := syslogTag(ev)
	message := syslogMessage(ev)
	return []byte(fmt.Sprintf("<%d>%s %s %s: %s", priority, timestamp, hostname, tag, message)), nil
}

// RFC5424 encodes events as RFC5424 syslog messages.
type RFC5424 struct{}

func init() {
	Register("syslog-rfc5424", RFC5424{})
}

// Name returns the encoder name.
func (RFC5424) Name() string { return "syslog-rfc5424" }

// ContentType returns the MIME type of the encoded output.
func (RFC5424) ContentType() string { return "text/plain" }

// Encode formats the event as an RFC5424 message.
func (RFC5424) Encode(ev event.Event) ([]byte, error) {
	priority := syslogPriority(ev)
	timestamp := ev.EventTime.Format("2006-01-02T15:04:05.000000Z")
	hostname := syslogHostname(ev)
	procid := "-"
	if ev.FlowID != nil {
		procid = *ev.FlowID
	}
	message := syslogMessage(ev)
	return []byte(fmt.Sprintf(
		"<%d>1 %s %s %s %s %s - %s",
		priority, timestamp, hostname, ev.SourceID, procid, ev.EventType, message,
	)), nil
}

File: eventweave/encoders/syslog.py
"""Syslog RFC3164 and RFC5424 encoders."""

from __future__ import annotations

import json

from eventweave.core.event import Event
from eventweave.encoders.base import Encoder, EncodeResult
from eventweave.encoders.registry import encoder


def _priority(event: Event) -> int:
    facility = event.attributes.get("syslog_facility", 16)
    severity = event.attributes.get("syslog_severity", 6)
    return int(facility) * 8 + int(severity)


def _message(event: Event) -> str:
    return str(
        event.attributes.get(
            "message",
            json.dumps(event.attributes, default=str, ensure_ascii=False),
        )
    )


def _hostname(event: Event) -> str:
    return str(event.attributes.get("hostname", event.source_id))


def _tag(event: Event) -> str:
    return str(event.attributes.get("syslog_tag", event.source_id))


@encoder("syslog-rfc3164", content_type="text/plain")
class Rfc3164Encoder(Encoder):
    """Encode an event as an RFC3164-style syslog message."""

    name = "syslog-rfc3164"
    content_type = "text/plain"

    def encode(self, event: Event) -> EncodeResult:
        try:
            priority = _priority(event)
        except (TypeError, ValueError) as exc:
            return self._fail(f"invalid syslog priority: {exc}")

        timestamp = event.event_time.strftime("%b %d %H:%M:%S")
        hostname = _hostname(event)
        tag = _tag(event)
        message = _message(event)
        line = f"<{priority}>{timestamp} {hostname} {tag}: {message}"
        return self._ok(line)


@encoder("syslog-rfc5424", content_type="text/plain")
class Rfc5424Encoder(Encoder):
    """Encode an event as an RFC5424-style syslog message."""

    name = "syslog-rfc5424"
    content_type = "text/plain"

    def encode(self, event: Event) -> EncodeResult:
        try:
            priority = _priority(event)
        except (TypeError, ValueError) as exc:
            return self._fail(f"invalid syslog priority: {exc}")

        timestamp = event.event_time.strftime("%Y-%m-%dT%H:%M:%S.%fZ")
        hostname = _hostname(event)
        app_name = event.source_id
        procid = event.flow_id or "-"
        msgid = event.event_type
        message = _message(event)
        line = (
            f"<{priority}>1 {timestamp} {hostname} {app_name} "
            f"{procid} {msgid} - {message}"
        )
        return self._ok(line)
